torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -583,13 +607,23 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
       auto& variable = replica.variables[intra_bucket_index];
       const auto offset = replica.offsets[intra_bucket_index];
       const auto length = replica.lengths[intra_bucket_index];
+
+      const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
+      bool global_unused =
+        local_used_maps_[var_idx.replica_index][var_idx.variable_index]
+          .item<int>() == 0;
```

Comment:
Since allreduce of these tensors happens before bucket allreduce and all allreduce are with the same stream, it seems work.wait of the first bucket implies the readiness of these tensors? It may not be good to assume the implicit synchronization though.

Reply:
Done. Now wait at finalize_backward.